### ISSUE TITLE
Allow NS records for non-origin records

### ISSUE THREAD

**Issue by thinkl33t** [pwaring/dns-sync-zone#24]
We have a requirement to be able to set NS records for subdomains that will be externally managed - for example making `aws.domain.co.uk be` managed by route53

Currently this doesn't work,  because NS and SOA records are stripped.  To keep the dont-break-your-domain functionality could we filter out NS and SOA records at the root level, but not for subdomains?  Or possibly only filter out NS records that refer to mythic beasts?

**Comment by pwaring** [comment]
The other option I can think of is to say that if you have any NS records in your zone file, your existing NS records will be stripped. If you have none, the ones in place will be kept. So you either specify all of them or none of them. Effectively if you choose to add NS records, we assume you know what you're doing, and on your own head be it if you break the zone.

**Comment by thinkl33t** [comment]
I guess a `--dont-strip-ns-records` flag might work too?

**Comment by pwaring** [comment]
Yes, though I think `--strip-ns-records` would probably be safer as that requires you to explicitly enable the dangerous option which most people won't want. I'll see if I can trial that in the next few days with one of my less important domains.

**Comment by thinkl33t** [comment]
Ah, yes.

I agree that the 'safe' way is the correct default, though i was thinking not stripping from the file as opposed to not stripping from the domain.  Maybe its a bad name.

`--include-dangerous-records`

**Comment by pwaring** [comment]
Working on this now.

**Comment by thinkl33t** [comment]
I've tested this, and it works perfectly for our purposes :)

**Comment by pwaring** [comment]
I think I've found a bug either in the code logic or the API, using the following RFC zone file:

```
$ORIGIN dnstest.uk

@ 300 A 46.235.226.238
@ 300 NS ns1.mythic-beasts.com.
@ 300 NS ns2.mythic-beasts.com.
@ 300 NS ns3.mythic-beasts.com.

delegated 300 NS ns1.mythic-beasts.com.
```

Two things happen if I try and upload this using the following command:

`pipenv run python3 -- ./sync-zone.py --credentials-file ../mythic-dns-data/credentials.json --zone dnstest.uk --rfc-file ../mythic-dns-data/zones/dnstest.uk --include-dangerous-records`

Error 1:

```
* Warning: deleting dangerous record:
@ 86400 NS ns1.mythic-beasts.com.
* Warning: deleting dangerous record:
@ 86400 NS ns2.mythic-beasts.com.
DELETE @ 86400 NS ns1.mythic-beasts.com.
DELETE @ 86400 NS ns2.mythic-beasts.com.
DELETE @ 300 A 46.235.226.238
DELETE delegated 300 NS ns1.mythic-beasts.com.
ADD @ 300 A 46.235.226.238
ADD @ 300 NS ns1.mythic-beasts.com.
ADD @ 300 NS ns2.mythic-beasts.com.
ADD @ 300 NS ns3.mythic-beasts.com.
ADD delegated 300 NS ns1.mythic-beasts.com.
Traceback (most recent call last):
  File "./sync-zone.py", line 455, in <module>
    print("* Error:", sync_response.code, sync_response.reason)
AttributeError: 'Response' object has no attribute 'code'
```

If I comment out line 455, error 2:

```
* Warning: deleting dangerous record:
@ 86400 NS ns1.mythic-beasts.com.
* Warning: deleting dangerous record:
@ 86400 NS ns2.mythic-beasts.com.
DELETE @ 86400 NS ns1.mythic-beasts.com.
DELETE @ 86400 NS ns2.mythic-beasts.com.
DELETE @ 300 A 46.235.226.238
DELETE delegated 300 NS ns1.mythic-beasts.com.
ADD @ 300 A 46.235.226.238
ADD @ 300 NS ns1.mythic-beasts.com.
ADD @ 300 NS ns2.mythic-beasts.com.
ADD @ 300 NS ns3.mythic-beasts.com.
ADD delegated 300 NS ns1.mythic-beasts.com.
NDELETE @ 86400 NS ns1.mythic-beasts.com.: No such record
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
ERR Entire transaction aborted due to earlier errors
```

It looks like the API doesn't like me deleting `ns1.mythic-beasts.com.`, even though I'm going to add it back and it appears if I access the API directly using `httpie`:

`http --form POST https://dnsapi.mythic-beasts.com/ domain='dnstest.uk' password='mypass' command='LIST'`

**Comment by pwaring** [comment]
I've just tested deleting the `ns1.mythic-beasts.com.` via the API instead of the script and get:

```
http --form POST https://dnsapi.mythic-beasts.com/ domain='dnstest.uk' password='mypass' command='DELETE @ 86400 NS ns1.mythic-beasts.com.'
HTTP/1.1 400 Bad Request
Access-Control-Allow-Headers: origin, content-type
Access-Control-Allow-Methods: GET, POST
Access-Control-Allow-Origin: *
Connection: close
Content-Type: text/plain
Date: Tue, 10 Mar 2020 15:02:28 GMT
Server: Apache/2.4.25 (Debian)
Transfer-Encoding: chunked

NDELETE @ 86400 NS ns1.mythic-beasts.com.: No such record

```

**Comment by pwaring** [comment]
First error has been fixed by #31.

**Comment by pwaring** [comment]
Testing and discussion with @nowster shows that if I leave off the `--include-dangerous` parameter, everything works as expected and the `@` `NS` records aren't deleted (but the subdomain ones are synced as expected). Will close this issue for now and produce a release.
